File: 435.non-overlapping-intervals.py
#
# @lc app=leetcode id=435 lang=python3
#
# [435] Non-overlapping Intervals
#

# @lc code=start
class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        intervals.sort(key=lambda x: x[1])
        print(intervals)
        
        removed = 0
        prev_end = intervals[0][1]

        for i in range(1, len(intervals)):
            if prev_end > intervals[i][0]:
                removed += 1
            else:
                prev_end = intervals[i][1]
        
        return removed

            

# @lc code=end

File: greedy/435.non-overlapping-intervals.swift
/*
 * @lc app=leetcode id=435 lang=swift
 *
 * [435] Non-overlapping Intervals
 */

// @lc code=start
class Solution {
    func eraseOverlapIntervals(_ intervals: [[Int]]) -> Int {
        guard intervals.count > 1 else {
            return 0
        }

        var sortedIntervals = intervals.sorted { $0[1] < $1[1] }

        var end = sortedIntervals[0][1]
        var count = 0

        for interval in sortedIntervals[1...] {
            if interval[0] < end {
                count += 1
            } else {
                end = interval[1]
            }
        }
        return count
    }
}
// @lc code=end

